tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRepository.scala

```diff
@@ -0,0 +1,117 @@
+import com.linagora.tmail.james.jmap.publicAsset.ImageContentType.ImageContentType
+import com.linagora.tmail.james.jmap.publicAsset.MemoryPublicAssetRepository.PublicAssetMetadata
+import org.apache.james.blob.api.{BlobId, BlobStore, BucketName}
+import org.apache.james.core.Username
+import org.apache.james.jmap.api.model.IdentityId
+import org.apache.james.jmap.api.model.Size.Size
+import org.apache.james.util.ReactorUtils
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import scala.jdk.CollectionConverters._
+
+trait PublicAssetRepository {
+  def create(username: Username, creationRequest: PublicAssetCreationRequest): Publisher[PublicAssetStorage]
+
+  def update(username: Username, id: PublicAssetId, identityIds: Set[IdentityId]): Publisher[Void]
+
+  def remove(username: Username, id: PublicAssetId): Publisher[Void]
+
+  def revoke(username: Username): Publisher[Void]
+
+  def get(username: Username, ids: Set[PublicAssetId]): Publisher[PublicAssetStorage]
+
+  def get(username: Username, id: PublicAssetId): Publisher[PublicAssetStorage] = get(username, Set(id))
+
+  def list(username: Username): Publisher[PublicAssetStorage]
+
+}
+
+object MemoryPublicAssetRepository {
+
+  object PublicAssetMetadata {
+    def from(publicAsset: PublicAssetStorage): PublicAssetMetadata =
+      PublicAssetMetadata(
+        publicAsset.id,
+        publicAsset.publicURI,
+        publicAsset.size,
+        publicAsset.contentType,
+        publicAsset.blobId,
+        publicAsset.identityIds)
+  }
+
+  case class PublicAssetMetadata(id: PublicAssetId,
+                                 publicURI: PublicURI,
+                                 size: Size,
+                                 contentType: ImageContentType,
+                                 blobId: BlobId,
+                                 identityIds: Seq[IdentityId]) {
+
+    def asPublicAssetStorage(bytes: Array[Byte]): PublicAssetStorage =
+      PublicAssetStorage(id = id,
+        publicURI = publicURI,
+        size = size,
+        contentType = contentType,
+        blobId = blobId,
+        identityIds = identityIds,
+        content = () => new ByteArrayInputStream(bytes))
+  }
+}
+
+class MemoryPublicAssetRepository(val blobStore: BlobStore) extends PublicAssetRepository {
+  private val tableStore: Table[Username, PublicAssetId, PublicAssetMetadata] = Tables.synchronizedTable(HashBasedTable.create())
+
+  private val bucketName: BucketName = blobStore.getDefaultBucketName
+
+  override def create(username: Username, creationRequest: PublicAssetCreationRequest): SMono[PublicAssetStorage] =
+    SMono.fromCallable(() => creationRequest.content.apply().readAllBytes())
+      .flatMap((dataAsByte: Array[Byte]) =>
+        SMono.fromCallable(() => new CountingInputStream(new ByteArrayInputStream(dataAsByte)))
+          .flatMap((dataAsByte: InputStream) => SMono(blobStore.save(bucketName, dataAsByte, BlobStore.StoragePolicy.LOW_COST)))
```

Comment:
If we have byte[] available I prefer we persist `byte[]` onto the blobStore rather than passing by the `InputStream` representation.